File: LeetCode_go/56.go
package main

import "sort"

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

func merge(intervals [][]int) [][]int {
	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i][0] < intervals[j][0]
	})

	var ans [][]int
	if len(intervals) == 0 {
		return ans
	}

	begin, end := intervals[0][0], intervals[0][1]
	for i := range intervals {
		tb, te := intervals[i][0], intervals[i][1]
		if tb <= end {
			begin = min(begin, tb)
			end = max(end, te)
		} else {
			ans = append(ans, []int{begin, end})
			begin, end = tb, te
		}
	}
	ans = append(ans, []int{begin, end})

	return ans
}

File: LeetCode/56.cpp
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        vector<vector<int>> ans;
        if (intervals.empty()) {
            return ans;
        }
        sort(intervals.begin(), intervals.end(), [](vector<int> x, vector<int> y) -> bool {
            return x[0] < y[0];
        });

        int begin = intervals[0][0];
        int end = intervals[0][1];
        for (int i = 1; i < intervals.size(); i++) {
            if (intervals[i][0] > end) {
                ans.push_back(vector<int>{begin, end});
                begin = intervals[i][0];
                end = intervals[i][1];
                continue;
            }
            end = max(end, intervals[i][1]);
        }
        ans.push_back(vector<int>{begin, end});
        return ans;
    }
};